lib/axon/loop.ex

```diff
@@ -1052,23 +1044,32 @@ defmodule Axon.Loop do
           {:halt, {:halt_loop, state}}
 
         {:continue, state} ->
-          batch_fn = build_batch_fn(step_fn, metric_fns)
-
-          %State{iteration: iters, max_iteration: max_iters} =
-            new_state = maybe_jit(batch_fn, [data, state], jit_compile?, jit_opts)
-
-          case fire_event(:iteration_completed, handler_fns, new_state) do
+          %State{
+            iteration: iters,
+            max_iteration: max_iters,
+            step_state: step_state,
+            metrics: metrics
+          } = state
+
+          {new_step_state, new_metrics} =
+            maybe_jit(batch_fn, [data, iters, step_state, metrics], jit_compile?, jit_opts)
+
+          state = %{
+            state
+            | iteration: iters + 1,
+              step_state: new_step_state,
+              metrics: new_metrics
+          }
```

Comment:
@seanmor5 now it all works! I found out the simplest approach is actually to not pass the state into defn/jit. Only the parts that we need and then get them back. This way there is no confusion about the boundaries and about what goes and what comes back.